pkg/syncer/mutators/deployment.go

```diff
@@ -0,0 +1,247 @@
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "deployments",
+	}
+}
+
+func (dt *DeploymentMutator) Register(mutators map[schema.GroupVersionResource]Mutator) {
+	if _, ok := mutators[dt.getGVR()]; !ok {
+		mutators[dt.getGVR()] = dt
+	}
+}
+
+func NewDeploymentMutator(fromConfig *rest.Config, kcpHost string) *DeploymentMutator {
+	return &DeploymentMutator{
+		fromConfig: fromConfig,
+		kcpHost:    kcpHost,
+	}
+}
+
+// ApplySpec makes modifications to the Spec of the deployment object.
+func (dt *DeploymentMutator) Mutate(downstreamObj *unstructured.Unstructured) error {
+	var deployment appsv1.Deployment
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		downstreamObj.UnstructuredContent(),
+		&deployment)
+	if err != nil {
+		return err
+	}
+
+	// If the deployment has not serviceaccount defined or it is "default", that means that it is using the default one
+	// and we need will need to override it with our own, "kcp-default"
+	//
+	// If the deployment has a serviceaccount defined, that means we need to synchronize the created service account
+	// down to the workloadcluster, so we don't modify it, and expect the scheduler to do the job.
+	//
+	if deployment.Spec.Template.Spec.ServiceAccountName == "" || deployment.Spec.Template.Spec.ServiceAccountName == "default" {
+		deployment.Spec.Template.Spec.ServiceAccountName = "kcp-default"
+	}
+
+	// Setting AutomountServiceAccountToken to false allow us to control the ServiceAccount
+	// VolumeMount and Volume definitions.
+	deployment.Spec.Template.Spec.AutomountServiceAccountToken = utilspointer.BoolPtr(false)
+
+	var hostname, port string
+	if dt.kcpHost != "" {
+		hostname, port, err = net.SplitHostPort(dt.kcpHost)
+		if err != nil {
+			return err
+		}
+	} else {
+		u, err := url.Parse(dt.fromConfig.Host)
+		if err != nil {
+			return err
+		}
+		hostname = u.Hostname()
+		port = u.Port()
+	}
+
+	kcpExternalPort := port
+	kcpExternalHost := hostname
+
+	overrideEnvs := []corev1.EnvVar{
+		{Name: "KUBERNETES_SERVICE_PORT", Value: kcpExternalPort},
+		{Name: "KUBERNETES_SERVICE_PORT_HTTPS", Value: kcpExternalPort},
+		{Name: "KUBERNETES_SERVICE_HOST", Value: kcpExternalHost},
+	}
+
+	// This is the VolumeMount that we will append to all the containers of the deployment
+	serviceAccountMount := corev1.VolumeMount{
+		Name:      "kcp-api-access",
+		MountPath: "/var/run/secrets/kubernetes.io/serviceaccount",
+		ReadOnly:  true,
+	}
+
+	// This is the Volume that we will add to the Deployment in order to control
+	// the name of the ca.crt references (kcp-root-ca.crt vs kube-root-ca.crt)
+	// and the serviceaccount reference.
+	serviceAccountVolume := corev1.Volume{
+		Name: "kcp-api-access",
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				DefaultMode: utilspointer.Int32Ptr(420),
+				Sources: []corev1.VolumeProjection{
+					{
+						// TODO(jmprusi): Instead of using a secret directly we should use the serviceaccount
```

Comment:
Do we actually want to do this? 

Bound tokens need to be refreshed periodically, which is something that the kubelet is guaranteed sufficient permission to do on behalf of a pod. Using that same model for a kcp-dependent pod would require a syncer component to have the same reliability of the kubelet. Except, if a kubelet fails to do its job, the pod goes unhealthy and is restarted somewhere else. If a syncer stops running for whatever reason, would it be preferable to allow the workload to keep running or have it die if its token isn't refreshed in a timely manner?